context/app/static/js/components/bulkDownload/hooks.ts

```diff
@@ -0,0 +1,199 @@
+import { useCallback, useEffect } from 'react';
+import { useForm } from 'react-hook-form';
+import { zodResolver } from '@hookform/resolvers/zod';
+import { z } from 'zod';
+
+import { bulkDownloadOptionsField, bulkDownloadMetadataField } from 'js/components/bulkDownload/bulkDownloadFormFields';
+import useBulkDownloadToasts from 'js/components/bulkDownload/toastHooks';
+import { ALL_BULK_DOWNLOAD_OPTIONS } from 'js/components/bulkDownload/constants';
+import { BulkDownloadDataset, useBulkDownloadStore } from 'js/stores/useBulkDownloadStore';
+import { useSearchHits } from 'js/hooks/useSearchData';
+import { useProtectedDatasetsForm } from 'js/hooks/useProtectedDatasets';
+import { createDownloadUrl } from 'js/helpers/functions';
+import { checkAndDownloadFile, postAndDownloadFile } from 'js/helpers/download';
+import { trackEvent } from 'js/helpers/trackers';
+import { getIDsQuery } from 'js/helpers/queries';
+
+const schema = z
+  .object({
+    ...bulkDownloadOptionsField,
+    ...bulkDownloadMetadataField,
+  })
+  .partial()
+  .required({ bulkDownloadOptions: true });
+
+export interface BulkDownloadFormTypes {
+  bulkDownloadOptions: string[];
+  bulkDownloadMetadata: boolean;
+}
+
+function useBulkDownloadForm() {
+  const {
+    handleSubmit,
+    control,
+    reset,
+    setValue,
+    formState: { errors },
+    trigger,
+  } = useForm<BulkDownloadFormTypes>({
+    defaultValues: {
+      bulkDownloadOptions: [],
+      bulkDownloadMetadata: false,
+    },
+    mode: 'onChange',
+    resolver: zodResolver(schema),
+  });
+
+  return {
+    handleSubmit,
+    setValue,
+    control,
+    errors,
+    reset,
+    trigger,
+  };
+}
+
+interface BulkDownloadDialogProps {
+  deselectRows?: (uuids: string[]) => void;
+}
+function useBulkDownloadDialog({ deselectRows }: BulkDownloadDialogProps) {
+  const { isOpen, uuids, open, close, setUuids } = useBulkDownloadStore();
+  const { control, handleSubmit, errors, reset, trigger } = useBulkDownloadForm();
+  const { toastErrorDownloadFile, toastSuccessDownloadFile } = useBulkDownloadToasts();
+
+  // Fetch datasets for the selected uuids
+  const datasetQuery = {
+    query: getIDsQuery([...uuids]),
+    _source: ['hubmap_id', 'processing', 'uuid', 'files', 'processing_type'],
+    size: 1000,
+  };
```

Comment:
It seems potentially worthwhile to put this into a `useMemo` to avoid recalculating it unless the uuids change.

This raised some other questions for me: 

- Is there an upper bound to how many datasets can be selected for bulk download at once? 
	- If so, are we communicating this to users? 
- If the limit is higher than 1,000, could we increase the size to `10_000` to handle selections of >1,000 datasets? We could also use `size: uuids.size()`.

Reply:
This is a good point and I like the suggestion of using `size: uuids.size()` - in regard to an upper bound, we haven't discussed this as far as I know. The load time for the dialog is definitely noticeable after a few hundred datasets, so it seems like some additional messaging should be provided to users who have selected a large number of datasets (maybe an alert similar to the one used in the workspaces dialog) letting them know about the wait time. @tsliaw what do you think?

Reply:
@john-conroy It's server side - once the search hits are returned the dialog loads almost immediately.

Reply:
@tsliaw @john-conroy I tried timing the wait time and it varies pretty substantially depending on what filters are applied - I'm guessing this is because of how the datasets are indexed?

For filters that include ~1000 datasets, the wait time is < 2 seconds for filters on dataset type, status, and analyte class, ~30 seconds for filters on sample category, and just over two minutes for filters on organs.